### PR TITLE
feat: improvements & gateway-conformance support (light)

### LINKED ISSUE #30: feat: add gateway-conformance tests
similar to https://github.com/ipfs/bifrost-gateway/blob/70f10e7b9688cd504f881d57ff3fa71215813e16/.github/workflows/gateway-conformance.yml

### Currently passing tests on https://github.com/ipfs/helia-http-gateway/pull/57 (last updated 2023-12-04):

<details>
  <summary>`subdomain-ipns-gateway`</summary>

```
PASS: TestMetadata (0.00s)
PASS: TestGatewaySubdomainAndIPNS/request_for_%7BCIDv1-libp2p-key%7D.ipns.%7Bgateway%7D_returns_expected_payload_%28direct_HTTP%29/Status_code (0.00s)
PASS: TestGatewaySubdomainAndIPNS/request_for_%7BCIDv1-libp2p-key%7D.ipns.%7Bgateway%7D_returns_expected_payload_%28HTTP_proxy%29/Status_code (0.00s)
PASS: TestGatewaySubdomainAndIPNS/request_for_%7BCIDv1-libp2p-key%7D.ipns.%7Bgateway%7D_returns_expected_payload_%28direct_HTTP%29#01/Status_code (0.00s)
PASS: TestGatewaySubdomainAndIPNS/request_for_%7BCIDv1-libp2p-key%7D.ipns.%7Bgateway%7D_returns_expected_payload_%28HTTP_proxy%29#01/Status_code (0.00s)
PASS: TestGatewaySubdomainAndIPNS/request_for_%7BCIDv1-libp2p-key%7D.ipns.%7Bgateway%7D_returns_expected_payload_%28direct_HTTP%29#02/Status_code (0.00s)
PASS: TestGatewaySubdomainAndIPNS/request_for_%7BCIDv1-libp2p-key%7D.ipns.%7Bgateway%7D_returns_expected_payload_%28HTTP_proxy%29#02/Status_code (0.00s)
PASS: TestGatewaySubdomainAndIPNS/request_for_%7BCIDv1-libp2p-key%7D.ipns.%7Bgateway%7D_returns_expected_payload_%28direct_HTTP%29#03/Status_code (0.00s)
PASS: TestGatewaySubdomainAndIPNS/request_for_%7BCIDv1-libp2p-key%7D.ipns.%7Bgateway%7D_returns_expected_payload_%28HTTP_proxy%29#03/Status_code (0.00s)
```
</details>

<details>
  <summary>`subdomain-ipfs-gateway`</summary>

```
PASS: TestMetadata (0.00s)
PASS: TestGatewaySubdomains/request_for_%7BCID%7D.ipfs.example.com_should_return_expected_payload_%28direct_HTTP%29/Status_code (0.00s)
PASS: TestGatewaySubdomains/request_for_%7BCID%7D.ipfs.example.com_should_return_expected_payload_%28direct_HTTP%29/Body (0.00s)
PASS: TestGatewaySubdomains/request_for_%7BCID%7D.ipfs.example.com_should_return_expected_payload_%28direct_HTTP%29 (0.99s)
PASS: TestGatewaySubdomains/request_for_%7BCID%7D.ipfs.example.com_should_return_expected_payload_%28HTTP_proxy%29/Status_code (0.00s)
PASS: TestGatewaySubdomains/valid_file_and_subdirectory_paths_in_directory_listing_at_%7Bcid%7D.ipfs.example.com_%28direct_HTTP%29/Status_code (0.00s)
PASS: TestGatewaySubdomains/valid_file_and_subdirectory_paths_in_directory_listing_at_%7Bcid%7D.ipfs.example.com_%28HTTP_proxy%29/Status_code (0.00s)
PASS: TestGatewaySubdomains/request_for_%7BCID%7D.ipfs.example.com_should_return_expected_payload_%28direct_HTTP%29#01/Status_code (0.00s)
PASS: TestGatewaySubdomains/request_for_%7BCID%7D.ipfs.example.com_should_return_expected_payload_%28direct_HTTP%29#01/Body (0.00s)
PASS: TestGatewaySubdomains/request_for_%7BCID%7D.ipfs.example.com_should_return_expected_payload_%28direct_HTTP%29#01 (0.00s)
PASS: TestGatewaySubdomains/request_for_%7BCID%7D.ipfs.example.com_should_return_expected_payload_%28HTTP_proxy%29#01/Status_code (0.00s)
PASS: TestGatewaySubdomains/valid_file_and_subdirectory_paths_in_directory_listing_at_%7Bcid%7D.ipfs.example.com_%28direct_HTTP%29#01/Status_code (0.00s)
PASS: TestGatewaySubdomains/valid_file_and_subdirectory_paths_in_directory_listing_at_%7Bcid%7D.ipfs.example.com_%28HTTP_proxy%29#01/Status_code (0.00s)
```
</details>

```[tasklist]
### Tasks
- [x] Figure out how to load fixtures (need to do via new api-endpoint if we're using in-memory block/data stores)
- [ ] Address failure for `http://localhost:8090/ipns/?uri=ipns%3A%2F%2Fdnslink-subdomain-gw-test.example.org` requests
- [ ] fix `http://localhost:8090/ipns/dnslink-subdomain-gw-test.example.org/wiki`
- [ ] fix `http://k51qzi5uqu5dk3v4rmjber23h16xnr23bsggmqqil9z2gduiis5se8dht36dam.ipns.localhost:8090
- [ ] fix `http://localhost:8090/ipns/QmVujd5Vb7moysJj8itnGufN7MEtPRCNHkKpNuA4onsRa3`
- [ ] fix IPNS resolution using IPNS fixtures loaded in kubo
```

### PR DESCRIPTION
**Pull request by aschmahmann**
This certainly isn't done yet (and not just because we don't pass the conformance tests yet). For example, the environment variables to help enable subdomain gateway tests are not wired up yet. Wanted to see how close this approach was to working though.

Closes #30 

Note: In addition to many tests failing, a number of them aren't showing up because some of the failures are killing the gateway-conformance tester due to this unresolved issue https://github.com/ipfs/gateway-conformance/blob/d855ec4fb9dac4a5aaecf3776037b005cc74c566/tooling/check/check.go#L358

### CODE REVIEW

File: .github/workflows/gateway-conformance.yml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+      # 6. Run the gateway-conformance tests
+      - name: Run gateway-conformance tests
+        uses: ipfs/gateway-conformance/.github/actions/test@v0.4.1
+        with:
+          gateway-url: http://127.0.0.1:8090
+          json: output.json
+          xml: output.xml
+          html: output.html
+          markdown: output.md
+          specs: +trustless-gateway,+path-gateway,+subdomain-gateway,+dnslink-gateway,+redirects-file,-path-dag-gateway
```
Comment (lidel) [inline .github/workflows/gateway-conformance.yml:97]: sounds like we want to descope trustless responses for SW work (SW will be mainly used for deserialized, trusted ones):

```suggestion
          specs: -trustless-gateway,+path-gateway,+subdomain-gateway,+dnslink-gateway,+redirects-file,-path-dag-gateway
```

**Comment by SgtPooki** [comment]
I don't think the gateway conformance test proxying works well with helia-http-gateway. something is up.. The test for url `http://localhost:8090/ipns/QmVujd5Vb7moysJj8itnGufN7MEtPRCNHkKpNuA4onsRa3`, titled "TestGatewaySubdomainAndIPNS/request_for_%2Fipns%2F%7BCIDv0%7D_redirects_to_CIDv1_with_libp2p-key_multicodec_in_subdomain_%28HTTP_proxy%29", says that it fails, `proxyconnect tcp: dial tcp [::1]:8090: connect: connection refused`, but when I test`http://localhost:8090/ipns/QmVujd5Vb7moysJj8itnGufN7MEtPRCNHkKpNuA4onsRa3` in my browser, it works fine. 

Is there a reason gateway-conformance is trying to use ipv6 addresses but reporting that it's trying to connect to `localhost:8090` ?

Also, a lot of requests are requesting `localhost/` instead of `localhost:port` and proxying them. this could be an issue with docker on my mac.. I will try without docker and see if I get similar errors.

**Comment by SgtPooki** [comment]
> Also, a lot of requests are requesting `localhost/` instead of `localhost:port` and proxying them. this could be an issue with docker on my mac.. I will try without docker and see if I get similar errors.


I confirmed that things are working much better running gateway-conformance directly with `go run ./cmd/gateway-conformance/main.go test --gateway-url 'http://localhost:8090' --subdomain-url 'http://localhost:8090' --json gwc-report.json --specs subdomain-ipns-gateway -- -timeout 30m`. I was able to get 9 passing tests this way.

**Comment by SgtPooki** [comment]
The report is [currently too big for an upload](https://github.com/ipfs/helia-http-gateway/actions/runs/7093356873/job/19306578893?pr=57#step:14:7), so I reduced the specs to just subdomain ipfs and ipns for now: 

`Error: $GITHUB_STEP_SUMMARY upload aborted, supports content up to a size of 1024k, got 1073k. For more information see: https://docs.github.com/actions/using-workflows/workflow-commands-for-github-actions#adding-a-markdown-summary`

**Comment by SgtPooki** [comment]
FYI: all gateway-conformance tests are disabled, but a lot of improvements are here that I want on the main branch. sorry folks.

**Pull request merged**